python/sglang/srt/distributed/device_communicators/pynccl_allocator.py

```diff
@@ -32,9 +31,22 @@
 
 
 def is_symmetric_memory_enabled():
+    # Import here to avoid circular import
+    from sglang.srt.managers.schedule_batch import global_server_args_dict
+
     return global_server_args_dict["enable_symm_mem"]
 
 
+def is_symmetric_memory_tensor(tensor: torch.Tensor):
+    if not is_symmetric_memory_enabled():
+        return False
+    for segment in get_nccl_mem_pool().snapshot():
```

Comment:
It's great that we no longer need to tag the tensors, but I'm concerned about the efficiency of this check. [Collecting the snapshot](https://github.com/pytorch/pytorch/blob/c5cb255625deb4cdbc5780e6911b73498e17ed5a/torch/csrc/cuda/Module.cpp#L724) seems to do a lot and then we also have to iterate over all segments/blocks.

1. Do you know if basic caching/memoization in this method would work using `tensor.untyped_storage.data_ptr()` as the key? Will the storage change from iteration to iteration or remain consistent?
2. If that doesn't work, could you measure how long this check takes?

Reply:
It would change, depending on what gets allocated in the pool But we could at least cache the snapshot at the exit of the context manager.

2. Yes I could time the call.

Reply:
133 us (not cached) vs 5us (cached)
So I made the changes to cache it at context manager exit.